sn_networking/src/replication_fetcher.rs

```diff
@@ -5,142 +5,195 @@
+use libp2p::{kad::RecordKey, PeerId};
 use rand::{seq::SliceRandom, thread_rng};
 use sn_protocol::NetworkAddress;
 use std::{
-    collections::{BTreeMap, HashSet},
+    collections::{BTreeMap, HashMap, HashSet},
     time::{Duration, Instant},
 };
 
-// Max parallel fetches can be undertaken at the same time.
+// Max parallel fetches that can be undertaken at the same time.
 const MAX_PARALLEL_FETCH: usize = 8;
 
 // The duration after which a peer will be considered failed to fetch data from,
 // if no response got from that peer.
-const FETCH_FAILED_DURATION: Duration = Duration::from_secs(10);
+const FETCH_TIMEOUT: Duration = Duration::from_secs(15);
+
+// The maximum number of retries that is performed per peer.
+// Else the key is fetched from the Network
+const MAX_RETRIES_PER_PEER: u8 = 1;
+
+// If we have failed to fetch the key from <= PEERS_TRIED_BEFORE_NETWORK_FETCH number of peers, then it is sent out
+// to be fetched from the network
+const PEERS_TRIED_BEFORE_NETWORK_FETCH: u8 = 3;
+
+// The number of failed attempts while fetching the key from a peer.
+type FailedAttempts = u8;
+// The time at which the key was sent to be fetched from the peer.
+type ReplicationRequestSentTime = Instant;
 
 // Status of the data fetching progress from the holder.
-#[derive(PartialEq)]
+#[derive(PartialEq, Debug)]
 pub(crate) enum HolderStatus {
     Pending,
     OnGoing,
-    Failed,
 }
 
 #[derive(Default)]
 pub(crate) struct ReplicationFetcher {
-    to_be_fetched: BTreeMap<NetworkAddress, BTreeMap<PeerId, (Instant, HolderStatus)>>,
+    to_be_fetched: HashMap<
+        RecordKey,
+        BTreeMap<PeerId, (ReplicationRequestSentTime, HolderStatus, FailedAttempts)>,
+    >,
     on_going_fetches: usize,
 }
 
 impl ReplicationFetcher {
-    /// Add a list of keys of a holder.
-    /// Return with a list of keys to fetch from holder
-    pub(crate) fn add_keys_to_replicate_per_peer(
+    // Adds the non existing incoming keys from the peer to the fetcher. Returns the next set of keys that has to be
+    // fetched from the peer/network.
+    pub(crate) fn add_keys(
         &mut self,
         peer_id: PeerId,
-        keys: Vec<NetworkAddress>,
-    ) -> Vec<(PeerId, NetworkAddress)> {
-        for key in keys {
-            self.add_holder(key, peer_id);
-        }
-        self.next_to_fetch()
-    }
-    /// Remove keys that we hold already and no longer need to be replicated.
-    pub(crate) fn remove_held_data(&mut self, keys: &HashSet<NetworkAddress>) {
-        self.to_be_fetched.retain(|key, _| !keys.contains(key));
+        incoming_keys: Vec<NetworkAddress>,
+        locally_stored_keys: &HashSet<RecordKey>,
+    ) -> Vec<(RecordKey, Option<PeerId>)> {
+        self.retain_keys(locally_stored_keys);
+
+        // add non existing keys to the fetcher
+        incoming_keys
+            .into_iter()
+            .filter_map(|incoming| incoming.as_record_key())
+            .filter(|incoming| !locally_stored_keys.contains(incoming))
```

Comment:
these two `filter_map` and `filter` shall be merged into just one `filter_map`

Reply:
That's possibly more optimised indeed but not sure it will increase readability?

Reply:
That's possibly more optimised indeed but not sure it will increase readability?